### DISCUSSION TITLE
Problem Creating a New Java Rule in PMD

### DISCUSSION THREAD

**Discussion by ShreyChandra** [pmd/pmd#3463] [Q&A] [answered]
Hi,

Apologies for asking something basic but I have been struggling to understand the documentation on how to add a new java rule.

I have been stuck on this for the past 3 days and now I can't sleep cause I cannot figure out how or even what should I do :(

I downloaded pmd-pmd_src-6.37.0 did a ./mvnw clean verify and got Build Success

I am mainly facing problem with the four "Basic" instructions mentioned in the docs, Please let me know if my understanding is wrong:

**a)write a Java class that implements the interface Rule. Each language implementation provides a base rule class to ease your pain, e.g. AbstractJavaRule.**

_For now am following the example of WhileLoopsMustUseBracesRule (https://pmd.sourceforge.io/pmd-5.3.2/customizing/howtowritearule.html) Just simply added the class to "../pmd-src-6.37.0/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/WhileLoopsMustUseBracesRule.java"_ and added my ruleset.xml at "../pmd-src-6.37.0/pmd-java/src/main/resources/rulesets/java/mycustomrules.xml"

**b)compile this class, linking it to PMD APIs (eg using PMD as a maven dependency)**

okay so I tried building the whole pmd-java module as well as did what Andreas told at https://sourceforge.net/p/pmd/discussion/188192/thread/7a34d224/ using command
javac -cp PATH_TO_PMD/lib/pmd-4.2.5.jar WhileLoopsMustUseBracesRule.java

I first tried to use the original path for java class file and ruleset path(mentioned above), and then I also tried placing everything in the /pmd-dist/target/pmd-bin-6.37.0/ 

my command - javac -cp lib/pmd-java.jar WhileLoopsMustUseBracesRule.java
pmd-java.jar was the jar that I created using WhileLoopMustUseBracesRule class.

**c)bundle this into a JAR and add it to the execution classpath of PMD**
as mentioned above I have already built the jar and added it to "../pmd-src-6.37.0/pmd-dist/target/pmd-bin-6.37.0/lib/pmd-java.jar". I hope that is what was meant by "add it to the execution classpath of PMD"

I don't suspect that I made a mistake while making the Jar cause my command - javac -cp lib/pmd-java.jar WhileLoopsMustUseBracesRule.java worked fine

**d)declare the rule in your ruleset XML**
Already mentioned above under point one.

Now when I run pmd using the command using the absolute paths for Example.java( has the violation code) and mycustomrule.xml I an error.

I tried moving the files to "../pmd-src-6.37.0" and ran the command:
bin/run.sh pmd -dir Example.java -f xml -rulesets mycustomrules.xml
or
bin/run.sh pmd -dir / -f csv -rulesets mycustomrules.xml
 
But I still get the error:
------------------------------------Error Log-------------------------------------
Aug 15, 2021 9:08:25 PM net.sourceforge.pmd.rules.RuleFactory buildRule
SEVERE: Error instantiating a rule
java.lang.ClassNotFoundException: WhileLoopsMustUseBracesRule
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:602)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	at net.sourceforge.pmd.util.ResourceLoader.loadRuleFromClassPath(ResourceLoader.java:139)
	at net.sourceforge.pmd.rules.RuleBuilder.build(RuleBuilder.java:196)
	at net.sourceforge.pmd.rules.RuleFactory.buildRule(RuleFactory.java:211)
	at net.sourceforge.pmd.RuleSetFactory.parseSingleRuleNode(RuleSetFactory.java:612)
	at net.sourceforge.pmd.RuleSetFactory.parseRuleNode(RuleSetFactory.java:501)
	at net.sourceforge.pmd.RuleSetFactory.parseRuleSetNode(RuleSetFactory.java:408)
	at net.sourceforge.pmd.RuleSetFactory.createRuleSet(RuleSetFactory.java:233)
	at net.sourceforge.pmd.RuleSetFactory.createRuleSet(RuleSetFactory.java:228)
	at net.sourceforge.pmd.RuleSetFactory.createRuleSets(RuleSetFactory.java:183)
	at net.sourceforge.pmd.RuleSetFactory.createRuleSets(RuleSetFactory.java:164)
	at net.sourceforge.pmd.RulesetsFactoryUtils.getRuleSets(RulesetsFactoryUtils.java:46)
	at net.sourceforge.pmd.RulesetsFactoryUtils.getRuleSetsWithBenchmark(RulesetsFactoryUtils.java:78)
	at net.sourceforge.pmd.PMD.doPMD(PMD.java:228)
	at net.sourceforge.pmd.PMD.runPmd(PMD.java:575)
	at net.sourceforge.pmd.PMD.runPmd(PMD.java:547)
	at net.sourceforge.pmd.PMD.main(PMD.java:503)

Mandatory arguments:
1) A java source code filename or directory
2) A report format 
3) A ruleset filename or a comma-delimited string of ruleset filenames

For example: 
C:\>pmd-bin-6.37.0\bin\pmd.bat -d c:\my\source\code -f html -R java-unusedcode

Languages and version suported:
apex, ecmascript, java, jsp, modelica, plsql, pom, scala, vf, vm, wsdl, xml, xsl

Available report formats and their configuration properties are:
   codeclimate: Code Climate integration.
   csv: Comma-separated values tabular format.
        problem - Include Problem column   default: true
        package - Include Package column   default: true
        file - Include File column   default: true
        priority - Include Priority column   default: true
        line - Include Line column   default: true
        desc - Include Description column   default: true
        ruleSet - Include Rule set column   default: true
        rule - Include Rule column   default: true
   emacs: GNU Emacs integration.
   empty: Empty, nothing.
   html: HTML format
        linePrefix - Prefix for line number anchor in the source file.
        linkPrefix - Path to HTML source.
        htmlExtension - Replace file extension with .html for the links (default: false)   default: false
   ideaj: IntelliJ IDEA integration.
        classAndMethodName - Class and Method name, pass '.method' when processing a directory.   default: 
        sourcePath - Source path.   default: 
        fileName - File name.   default: 
   json: JSON format.
   sarif: Static Analysis Results Interchange Format (SARIF)
   summaryhtml: Summary HTML format.
        linePrefix - Prefix for line number anchor in the source file.
        linkPrefix - Path to HTML source.
        htmlExtension - Replace file extension with .html for the links (default: false)   default: false
   text: Text format.
   textcolor: Text format, with color support (requires ANSI console support, e.g. xterm, rxvt, etc.).
        color - Enables colors with anything other than 'false' or '0'.   default: yes
   textpad: TextPad integration.
   vbhtml: Vladimir Bossicard HTML format.
   xml: XML format.
        encoding - XML encoding format, defaults to UTF-8.   default: UTF-8
   xslt: XML with a XSL Transformation applied.
        encoding - XML encoding format, defaults to UTF-8.   default: UTF-8
        xsltFilename - The XSLT file name.
   yahtml: Yet Another HTML format.
        outputDir - Output directory.

For example on windows: 
C:\>pmd-bin-6.37.0\bin\pmd.bat -dir c:\my\source\code -format text -R rulesets/java/quickstart.xml -version 1.5 -language java -debug
C:\>pmd-bin-6.37.0\bin\pmd.bat -dir c:\my\source\code -f xml -rulesets rulesets/java/quickstart.xml,category/java/codestyle.xml -encoding UTF-8
C:\>pmd-bin-6.37.0\bin\pmd.bat -d c:\my\source\code -rulesets rulesets/java/quickstart.xml -auxclasspath lib\commons-collections.jar;lib\derby.jar
C:\>pmd-bin-6.37.0\bin\pmd.bat -d c:\my\source\code -f html -R rulesets/java/quickstart.xml -auxclasspath file:///C:/my/classpathfile

For example on *nix: 
$ pmd-bin-6.37.0/bin/run.sh pmd -dir /home/workspace/src/main/java/code -f html -rulesets rulesets/java/quickstart.xml,category/java/codestyle.xml
$ pmd-bin-6.37.0/bin/run.sh pmd -d ./src/main/java/code -R rulesets/java/quickstart.xml -f xslt -property xsltFilename=my-own.xsl
$ pmd-bin-6.37.0/bin/run.sh pmd -d ./src/main/java/code -f html -R rulesets/java/quickstart.xml -auxclasspath commons-collections.jar:derby.jar

java.lang.ClassNotFoundException: WhileLoopsMustUseBracesRule
--------------------------------------------------------------------------------------------

Thanks!

**Comment by adangel** [comment, marked as answer]
Just to point to the up to date documentation: https://pmd.github.io/latest/pmd_userdocs_extending_writing_rules_intro.html and https://pmd.github.io/latest/pmd_userdocs_extending_writing_java_rules.html#basics - the basic steps are the same (as in your old mentioned doc from version 5.3.2).

Let's go through the steps:

a) write a Java class that implements the interface Rule. Each language implementation provides a base rule class to ease your pain, e.g. AbstractJavaRule.

You have obviously written a file `WhileLoopsMustUseBracesRule.java`. Let's say, you store the file under `$HOME/pmd/customrule/WhileLoopsMustUseBraces.java`.

> Just simply added the class to "../pmd-src-6.37.0/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/WhileLoopsMustUseBracesRule.java" and added my ruleset.xml at "../pmd-src-6.37.0/pmd-java/src/main/resources/rulesets/java/mycustomrules.xml"

That's basically not necessary. You can develop your custom rules separately from PMD itself. You don't even need the source code package (But having the source helps you develop your rule in the IDE). Only when you eventually want to contribute the rule, you'll need the source code. But then, you have to fork the repository on github as otherwise you can't create a pull request.

b) compile this class, linking it to PMD APIs (eg using PMD as a maven dependency)

So, there are many different ways how to get PMD APIs to link it against. If you don't use a IDE or build tool (like maven), you can download the binary distribution from https://github.com/pmd/pmd/releases/latest (pmd-bin-6.37.0.zip), extract that, e.g. to `$HOME/pmd/`. Then you'll have a there a directory `$HOME/pmd/pmd-bin-6.37.0/lib` which contains all the jars you'll need to compile your custom rule:

    cd "$HOME/pmd/customrule"
    javac -cp "$HOME/pmd/pmd-bin-6.37.0/lib/*" WhileLoopsMustUseBracesRule.java

This creates the files `WhileLoopsMustUseBracesRule.class` in the same directory.

> my command - javac -cp lib/pmd-java.jar WhileLoopsMustUseBracesRule.java
> pmd-java.jar was the jar that I created using WhileLoopMustUseBracesRule class.

I think, that's the misunderstanding: the command javac (the java compiler) only *uses* pmd-java.jar but does *not create* it. It creates the class file WhileLoopMustUseBracesRule.class. But as you have seen, it is nowhere included in the jars and is missing afterwards. pmd-java.jar is not modified by javac.

c) bundle this into a JAR and add it to the execution classpath of PMD

You can't add single class files onto the classpath, you need to put them into jars and put the jar onto the classpath or alternatively you can put the directory, where the class file is, onto the classpath. But creating a jar is easy:

    cd "$HOME/pmd/customrule"
    jar --create --file customrule.jar WhileLoopsMustUseBracesRule.class

Now we have the file `$HOME/pmd/customrule/customrule.jar`.

This file needs to be added to the execution classpath. One simple way to do this, is, to copy jar this into the lib folder:

    cp "$HOME/pmd/customrule/customrule.jar" "$HOME/pmd/pmd-bin-6.37.0/lib/"

(Note: there is an alternative way, using the [environment variable CLASSPATH](https://github.com/pmd/pmd-examples/tree/java-without-maven#use-the-custom-rule-with-pmd-cli))

d) declare the rule in your ruleset XML

You need a ruleset xml file, e.g.

```xml
<?xml version="1.0"?>
<ruleset name="Custom Rules"
    xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">

    <description>My custom rules</description>

    <rule name="WhileLoopsMustUseBraces"
          language="java"
          message="While loops must use braces"
          class="WhileLoopsMustUseBracesRule" >
        <description>
TODO
        </description>
        <priority>3</priority>
    </rule>

</ruleset>
```

Store this as `$HOME/pmd/customrule/customrule.xml`

Now you can run PMD using your custom rule:

    $HOME/pmd/pmd-bin-6.37.0/bin/run.sh pmd -dir Example.java -f xml -rulesets $HOME/pmd/customrule/customrule.xml



Your error "ClassNotFoundException: WhileLoopsMustUseBracesRule" is probably due to the fact, that the class file "WhileLoopsMustUseBracesRule.class" is not on the execution classpath of PMD (step c)).

-----

You might also want to have a look at the following sample projects:

* [Java Custom Rules without maven](https://github.com/pmd/pmd-examples/tree/java-without-maven)
* [Java Custom Rules](https://github.com/pmd/pmd-examples/tree/java)

**Reply by ShreyChandra** [reply]
Thank you so much. Worked Like a charm. :1st_place_medal:

**Comment by vedant011** [comment]
![image](https://user-images.githubusercontent.com/64039474/202843436-c5fc31f6-cb66-464e-8f2e-74fb15fff9c8.png)
I have done all the steps mentioned above but still getting this error, please help me thanks in advance

**Reply by adangel** [reply]
The error message indicates, that the class "WhileLoopsMustUseBracesRule" is not on the classpath.

Please execute `zipinfo customrule.jar` and post the results here. It should contain a file "WhileLoopsMustUseBracesRule.class".

**Comment by vedant011** [comment]
![image](https://user-images.githubusercontent.com/64039474/202843882-81f38b88-538d-4b54-bb58-83292d4a5ff7.png)
i got this result what should i do next?

**Reply by adangel** [reply]
The problem is, that the class file is "lib/WhileLoopsMustUseBracesRule.class" and not "WhileLoopsMustUseBracesRule.class". The jar file has the wrong content. Which command did you use to create this jar file?

**Reply by vedant011** [reply]
jar cvf customrule.jar ../lib/WhileLoopsMustUseBracesRule.class
this one
how can i solve this problem?? kindly help me out

**Reply by adangel** [reply]
You need to switch directories before you create the jar (the jar command adds the files relative to the current directory):

```
cd ../lib
jar cvf ../customrule/customrule.jar WhileLoopsMustUseBracesRule.class
cd ../customrule
```

**Reply by vedant011** [reply]
Thanks alot it worked for me!!
How can we generate report file in html using intellij instead of command lines?

**Comment by vedant011** [comment]
For checking custom rules i'm changing the code for that particular rule so everytime i have to create the jar again is their any other method or way to do this?
It's very time consuming to create jar again and again.

**Reply by adangel** [reply]
See #4222
